File: bindings/javascript/packages/core/src/qaoa.ts
/**
 * Quantum Approximate Optimization Algorithm (C-side since v0.2.0;
 * JS binding since v0.5.5).
 *
 * Wraps ``src/algorithms/qaoa.{c,h}`` around the Ising-model
 * encoding, the MaxCut graph helper, and the ``qaoa_solve`` driver.
 * Mirrors the Python ``moonlab.algorithms.QAOA`` and Rust
 * ``moonlab::qaoa`` surfaces.
 *
 * @example
 * ```typescript
 * import { Graph, IsingModel, QaoaSolver } from '@moonlab/quantum-core';
 *
 * // Triangle MaxCut: optimal cut value = 2.
 * const g = await Graph.create(3, [[0, 1, 1.0], [1, 2, 1.0], [2, 0, 1.0]]);
 * const ising = await IsingModel.fromMaxcut(g);
 * const solver = await QaoaSolver.create(ising, 1);
 * const r = solver.solve();
 * console.log(`best energy = ${r.bestEnergy.toFixed(3)}`);
 * solver.dispose();
 * ```
 */

import { getModule } from './wasm-loader';

/** QAOA optimisation result.  Mirrors ``qaoa_result_t`` minus the
 *  raw pointers; `optimalGamma` / `optimalBeta` / `energyHistory`
 *  are copied into owned `Float64Array`s. */
export interface QaoaResult {
  /** Bitstring of the best-energy sample seen during optimisation. */
  bestBitstring: bigint;
  /** Energy at `bestBitstring`. */
  bestEnergy: number;
  /** Number of optimiser iterations consumed. */
  numIterations: number;
  /** `true` if the optimiser hit the convergence tolerance. */
  converged: boolean;
  /** Solution quality vs the global optimum. */
  approximationRatio: number;
  /** Optimal cost-Hamiltonian angles `gamma_1, ..., gamma_p`. */
  optimalGamma: Float64Array;
  /** Optimal mixer-Hamiltonian angles `beta_1, ..., beta_p`. */
  optimalBeta: Float64Array;
  /** QAOA depth `p`. */
  numLayers: number;
  /** Total shot count consumed across the run. */
  totalMeasurements: number;
}

type QaoaModule = {
  _graph_create: (numVertices: number, numEdges: number) => number;
  _graph_free: (g: number) => void;
  _graph_add_edge: (g: number, idx: number, u: number, v: number, w: number) => number;
  _ising_model_create: (numQubits: number) => number;
  _ising_model_free: (ising: number) => void;
  _ising_model_set_coupling: (ising: number, i: number, j: number, v: number) => number;
  _ising_model_set_field: (ising: number, i: number, v: number) => number;
  _ising_model_evaluate: (ising: number, bitstring: bigint) => number;
  _ising_encode_maxcut: (graph: number) => number;
  _qaoa_solver_create: (ising: number, numLayers: number, entropy: number) => number;
  _qaoa_solver_free: (solver: number) => void;
  _qaoa_solve: (resultPtr: number, solver: number) => void;
  _qaoa_compute_expectation: (solver: number, gamma: number, beta: number) => number;
  _quantum_entropy_ctx_create_hw: () => number;
  _quantum_entropy_ctx_destroy: (c: number) => void;
  _malloc: (n: number) => number;
  _free: (p: number) => void;
  HEAPF64: Float64Array;
  HEAP32: Int32Array;
  HEAPU32: Uint32Array;
};

/** Weighted undirected graph used to encode MaxCut. */
export class Graph {
  private ptr: number;
  private mod: QaoaModule;
  private freed = false;

  /** @internal */
  _internal_ptr(): number {
    if (this.freed) throw new Error('Graph disposed');
    return this.ptr;
  }

  private constructor(ptr: number, mod: QaoaModule) {
    this.ptr = ptr;
    this.mod = mod;
  }

  /** Build a graph on `numVertices` vertices with the given
   *  weighted edges `[u, v, weight]`. */
  static async create(
    numVertices: number,
    edges: ReadonlyArray<readonly [number, number, number]>,
  ): Promise<Graph> {
    const mod = (await getModule()) as unknown as QaoaModule;
    const ptr = mod._graph_create(numVertices, edges.length);
    if (!ptr) throw new Error('graph_create returned NULL');
    for (let i = 0; i < edges.length; i++) {
      const [u, v, w] = edges[i];
      const rc = mod._graph_add_edge(ptr, i, u, v, w);
      if (rc !== 0) {
        mod._graph_free(ptr);
        throw new Error(`graph_add_edge(idx=${i}, u=${u}, v=${v}) rc=${rc}`);
      }
    }
    return new Graph(ptr, mod);
  }

  /** Release the underlying allocation. */
  dispose(): void {
    if (this.freed) return;
    this.mod._graph_free(this.ptr);
    this.freed = true;
  }
}

/** Ising-model Hamiltonian
 *  `H = sum_{i<j} J_{ij} s_i s_j + sum_i h_i s_i`. */
export class IsingModel {
  private ptr: number;
  private mod: QaoaModule;
  private freed = false;

  /** @internal */
  _internal_ptr(): number {
    if (this.freed) throw new Error('IsingModel disposed');
    return this.ptr;
  }

  private constructor(ptr: number, mod: QaoaModule) {
    this.ptr = ptr;
    this.mod = mod;
  }

  /** Empty Ising model on `numQubits` spins.  Add couplings / fields
   *  via {@link setCoupling} / {@link setField}. */
  static async create(numQubits: number): Promise<IsingModel> {
    const mod = (await getModule()) as unknown as QaoaModule;
    const ptr = mod._ising_model_create(numQubits);
    if (!ptr) throw new Error('ising_model_create returned NULL');
    return new IsingModel(ptr, mod);
  }

  /** Encode a MaxCut problem on `graph` as the equivalent Ising
   *  model: `H = sum_{(i,j) in E} w_{ij} (1 - s_i s_j) / 2`. */
  static async fromMaxcut(graph: Graph): Promise<IsingModel> {
    const mod = (await getModule()) as unknown as QaoaModule;
    const ptr = mod._ising_encode_maxcut(graph._internal_ptr());
    if (!ptr) throw new Error('ising_encode_maxcut returned NULL');
    return new IsingModel(ptr, mod);
  }

  /** Set the symmetric coupling `J_{ij}`. */
  setCoupling(i: number, j: number, value: number): this {
    if (this.freed) throw new Error('IsingModel disposed');
    const rc = this.mod._ising_model_set_coupling(this.ptr, i, j, value);
    if (rc !== 0) throw new Error(`ising_model_set_coupling(${i},${j}) rc=${rc}`);
    return this;
  }

  /** Set the on-site field `h_i`. */
  setField(i: number, value: number): this {
    if (this.freed) throw new Error('IsingModel disposed');
    const rc = this.mod._ising_model_set_field(this.ptr, i, value);
    if (rc !== 0) throw new Error(`ising_model_set_field(${i}) rc=${rc}`);
    return this;
  }

  /** Evaluate the classical Ising energy on the bitstring
   *  `bits[i] = (bitstring >> i) & 1`. */
  evaluate(bitstring: bigint): number {
    if (this.freed) throw new Error('IsingModel disposed');
    return this.mod._ising_model_evaluate(this.ptr, bitstring);
  }

  /** Release the underlying allocation. */
  dispose(): void {
    if (this.freed) return;
    this.mod._ising_model_free(this.ptr);
    this.freed = true;
  }
}

/** QAOA solver bundling `(ising, numLayers, entropy)`. */
export class QaoaSolver {
  private solver: number;
  private entropy: number;
  private ising: IsingModel;
  private mod: QaoaModule;
  private freed = false;

  private constructor(
    solver: number, entropy: number, ising: IsingModel, mod: QaoaModule,
  ) {
    this.solver = solver;
    this.entropy = entropy;
    this.ising = ising;
    this.mod = mod;
  }

  /** Build a QAOA solver at depth `numLayers` over `ising`. */
  static async create(ising: IsingModel, numLayers: number): Promise<QaoaSolver> {
    if (numLayers < 1) throw new RangeError('numLayers must be >= 1');
    const mod = (await getModule()) as unknown as QaoaModule;
    const entropy = mod._quantum_entropy_ctx_create_hw();
    if (!entropy) throw new Error('quantum_entropy_ctx_create_hw returned NULL');
    const solver = mod._qaoa_solver_create(
      ising._internal_ptr(), numLayers, entropy,
    );
    if (!solver) {
      mod._quantum_entropy_ctx_destroy(entropy);
      throw new Error('qaoa_solver_create returned NULL');
    }
    return new QaoaSolver(solver, entropy, ising, mod);
  }

  /** Run the QAOA optimisation loop to convergence. */
  solve(): QaoaResult {
    if (this.freed) throw new Error('QaoaSolver disposed');
    // qaoa_result_t is 64 bytes on wasm32; emcc passes the struct
    // through a hidden first argument when returning by value.
    const resultPtr = this.mod._malloc(64);
    try {
      this.mod._qaoa_solve(resultPtr, this.solver);
      // Field offsets confirmed by C-side probe under emcc 4.0.22:
      //  0  uint64_t best_bitstring
      //  8  double   best_energy
      // 16  ptr      energy_history
      // 20  size_t   num_iterations
      // 24  int      converged
      // 32  double   approximation_ratio
      // 40  ptr      optimal_gamma
      // 44  ptr      optimal_beta
      // 48  size_t   num_layers
      // 52  size_t   total_measurements
      const bitLo = BigInt(this.mod.HEAPU32[resultPtr >> 2]);
      const bitHi = BigInt(this.mod.HEAPU32[(resultPtr + 4) >> 2]);
      const bestBitstring     = (bitHi << 32n) | bitLo;
      const bestEnergy        = this.mod.HEAPF64[(resultPtr + 8) >> 3];
      const numIterations     = this.mod.HEAPU32[(resultPtr + 20) >> 2];
      const converged         = this.mod.HEAP32[(resultPtr + 24) >> 2] !== 0;
      const approximationRatio = this.mod.HEAPF64[(resultPtr + 32) >> 3];
      const gammaPtr          = this.mod.HEAPU32[(resultPtr + 40) >> 2];
      const betaPtr           = this.mod.HEAPU32[(resultPtr + 44) >> 2];
      const numLayers         = this.mod.HEAPU32[(resultPtr + 48) >> 2];
      const totalMeasurements = this.mod.HEAPU32[(resultPtr + 52) >> 2];

      const copyOut = (ptr: number, n: number): Float64Array => (
        ptr && n > 0
          ? new Float64Array(this.mod.HEAPF64.buffer, ptr, n).slice()
          : new Float64Array(0)
      );
      return {
        bestBitstring,
        bestEnergy,
        numIterations,
        converged,
        approximationRatio,
        optimalGamma: copyOut(gammaPtr, numLayers),
        optimalBeta: copyOut(betaPtr, numLayers),
        numLayers,
        totalMeasurements,
      };
    } finally {
      this.mod._free(resultPtr);
    }
  }

  /** Evaluate `<H_C>(gamma, beta)` at the given angles without
   *  running the optimiser. */
  computeExpectation(gamma: Float64Array, beta: Float64Array): number {
    if (this.freed) throw new Error('QaoaSolver disposed');
    const gPtr = this.mod._malloc(gamma.length * 8);
    const bPtr = this.mod._malloc(beta.length * 8);
    try {
      for (let i = 0; i < gamma.length; i++) {
        this.mod.HEAPF64[(gPtr >> 3) + i] = gamma[i];
      }
      for (let i = 0; i < beta.length; i++) {
        this.mod.HEAPF64[(bPtr >> 3) + i] = beta[i];
      }
      return this.mod._qaoa_compute_expectation(this.solver, gPtr, bPtr);
    } finally {
      this.mod._free(gPtr);
      this.mod._free(bPtr);
    }
  }

  /** Release every C-side allocation owned by this solver. */
  dispose(): void {
    if (this.freed) return;
    this.mod._qaoa_solver_free(this.solver);
    this.mod._quantum_entropy_ctx_destroy(this.entropy);
    this.freed = true;
  }
}

File: bindings/rust/moonlab/src/qaoa.rs
//! Quantum Approximate Optimization Algorithm (since v0.2.0; safe
//! Rust wrapper since v0.4.9).
//!
//! Wraps `src/algorithms/qaoa.{c,h}` with idiomatic Rust around the
//! Ising-model encoding, the MaxCut graph helper, and the `qaoa_solve`
//! driver.  Mirrors the v0.2.0 Python `moonlab.algorithms.QAOA`
//! surface.
//!
//! [`QaoaSolver`] leases a hardware-entropy context for shot-noise
//! sampling via the same RAII pattern as [`crate::bell`] /
//! [`crate::grover`] / [`crate::vqe`].
//!
//! ## Quick start
//!
//! ```no_run
//! use moonlab::qaoa::{Graph, IsingModel, QaoaSolver};
//!
//! // Triangle graph -- MaxCut value = 2 (cut 2 of 3 edges).
//! let g = Graph::new(3, &[(0, 1, 1.0), (1, 2, 1.0), (2, 0, 1.0)]).unwrap();
//! let ising = IsingModel::from_maxcut(&g).unwrap();
//! let mut solver = QaoaSolver::new(ising, /*p=*/1).unwrap();
//! let result = solver.solve().unwrap();
//! println!("best cut value = {:.3}", -result.best_energy);
//! ```

use crate::error::{QuantumError, Result};
use moonlab_sys::{
    graph_add_edge, graph_create, graph_free, graph_t,
    ising_encode_maxcut, ising_model_create, ising_model_evaluate,
    ising_model_free, ising_model_set_coupling, ising_model_set_field,
    ising_model_t, qaoa_compute_expectation, qaoa_result_t, qaoa_solve,
    qaoa_solver_create, qaoa_solver_free, qaoa_solver_t,
    quantum_entropy_ctx_create_hw, quantum_entropy_ctx_destroy,
    quantum_entropy_ctx_t,
};
use std::ptr;

struct EntropyGuard {
    ctx: *mut quantum_entropy_ctx_t,
}

impl EntropyGuard {
    fn new() -> Result<Self> {
        let ctx = unsafe { quantum_entropy_ctx_create_hw() };
        if ctx.is_null() {
            return Err(QuantumError::Ffi(
                "quantum_entropy_ctx_create_hw returned NULL".to_string(),
            ));
        }
        Ok(Self { ctx })
    }
}

impl Drop for EntropyGuard {
    fn drop(&mut self) {
        if !self.ctx.is_null() {
            unsafe { quantum_entropy_ctx_destroy(self.ctx) };
            self.ctx = ptr::null_mut();
        }
    }
}

/// Weighted undirected graph used to encode MaxCut.
pub struct Graph {
    ptr: *mut graph_t,
}

impl Graph {
    /// Build a graph on `num_vertices` vertices with the given
    /// weighted edges `(u, v, weight)`.
    pub fn new(num_vertices: usize, edges: &[(usize, usize, f64)]) -> Result<Self> {
        let ptr = unsafe { graph_create(num_vertices, edges.len()) };
        if ptr.is_null() {
            return Err(QuantumError::Ffi(
                "graph_create returned NULL".to_string(),
            ));
        }
        for (idx, &(u, v, w)) in edges.iter().enumerate() {
            let rc = unsafe { graph_add_edge(ptr, idx, u as i32, v as i32, w) };
            if rc != 0 {
                unsafe { graph_free(ptr) };
                return Err(QuantumError::Ffi(format!(
                    "graph_add_edge(idx={idx}, u={u}, v={v}) rc={rc}"
                )));
            }
        }
        Ok(Self { ptr })
    }

    pub(crate) fn as_ptr(&self) -> *const graph_t {
        self.ptr
    }
}

impl Drop for Graph {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { graph_free(self.ptr) };
            self.ptr = ptr::null_mut();
        }
    }
}

/// Ising-model Hamiltonian `H = sum_{i<j} J_{ij} s_i s_j + sum_i h_i s_i`
/// over `num_qubits` spins.
pub struct IsingModel {
    ptr: *mut ising_model_t,
}

impl IsingModel {
    /// Empty Ising model on `num_qubits` spins.  Add couplings and
    /// fields via [`set_coupling`](Self::set_coupling) /
    /// [`set_field`](Self::set_field).
    pub fn new(num_qubits: usize) -> Result<Self> {
        let ptr = unsafe { ising_model_create(num_qubits) };
        if ptr.is_null() {
            return Err(QuantumError::Ffi(
                "ising_model_create returned NULL".to_string(),
            ));
        }
        Ok(Self { ptr })
    }

    /// Encode a MaxCut problem on `graph` as the equivalent Ising
    /// model: `H = sum_{(i,j) in E} w_{ij} (1 - s_i s_j) / 2`.
    pub fn from_maxcut(graph: &Graph) -> Result<Self> {
        let ptr = unsafe { ising_encode_maxcut(graph.as_ptr()) };
        if ptr.is_null() {
            return Err(QuantumError::Ffi(
                "ising_encode_maxcut returned NULL".to_string(),
            ));
        }
        Ok(Self { ptr })
    }

    /// Set the symmetric coupling `J_{ij}`.
    pub fn set_coupling(&mut self, i: usize, j: usize, value: f64) -> Result<()> {
        let rc = unsafe { ising_model_set_coupling(self.ptr, i, j, value) };
        if rc != 0 {
            return Err(QuantumError::Ffi(format!(
                "ising_model_set_coupling({i},{j}) rc={rc}"
            )));
        }
        Ok(())
    }

    /// Set the on-site field `h_i`.
    pub fn set_field(&mut self, i: usize, value: f64) -> Result<()> {
        let rc = unsafe { ising_model_set_field(self.ptr, i, value) };
        if rc != 0 {
            return Err(QuantumError::Ffi(format!(
                "ising_model_set_field({i}) rc={rc}"
            )));
        }
        Ok(())
    }

    /// Evaluate the classical Ising energy on the bitstring
    /// `bits[i] = (bitstring >> i) & 1`, where `s_i = 1 - 2 bits[i]`.
    pub fn evaluate(&self, bitstring: u64) -> f64 {
        unsafe { ising_model_evaluate(self.ptr, bitstring) }
    }
}

impl Drop for IsingModel {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { ising_model_free(self.ptr) };
            self.ptr = ptr::null_mut();
        }
    }
}

/// QAOA optimisation result.  Mirrors `qaoa_result_t` minus the raw
/// `energy_history` / `optimal_gamma` / `optimal_beta` pointers;
/// those are copied into owned `Vec<f64>`.
#[derive(Debug, Clone)]
pub struct QaoaResult {
    /// Bitstring of the best-energy sample seen during optimisation.
    pub best_bitstring: u64,
    /// Energy at `best_bitstring`.
    pub best_energy: f64,
    /// Number of optimiser iterations consumed.
    pub num_iterations: usize,
    /// `true` if the optimiser hit the convergence tolerance.
    pub converged: bool,
    /// Solution quality vs the global optimum (when known).
    pub approximation_ratio: f64,
    /// Optimal cost-Hamiltonian angles `gamma_1, ..., gamma_p`.
    pub optimal_gamma: Vec<f64>,
    /// Optimal mixer-Hamiltonian angles `beta_1, ..., beta_p`.
    pub optimal_beta: Vec<f64>,
    /// QAOA depth `p`.
    pub num_layers: usize,
    /// Total shot count consumed across the run.
    pub total_measurements: usize,
}

/// QAOA solver bundling `(ising, num_layers, entropy)`.
pub struct QaoaSolver {
    solver: *mut qaoa_solver_t,
    ising: IsingModel,
    _entropy: EntropyGuard,
}

impl QaoaSolver {
    /// Build a QAOA solver at depth `num_layers` over `ising`.
    /// `num_layers >= 1`.
    pub fn new(ising: IsingModel, num_layers: usize) -> Result<Self> {
        if num_layers == 0 {
            return Err(QuantumError::Ffi(
                "num_layers must be >= 1".to_string(),
            ));
        }
        let entropy = EntropyGuard::new()?;
        let solver = unsafe {
            qaoa_solver_create(ising.ptr, num_layers, entropy.ctx)
        };
        if solver.is_null() {
            return Err(QuantumError::Ffi(
                "qaoa_solver_create returned NULL".to_string(),
            ));
        }
        Ok(Self {
            solver,
            ising,
            _entropy: entropy,
        })
    }

    /// Number of spins in the underlying Ising problem.
    pub fn num_qubits(&self) -> usize {
        // qaoa_solver_t doesn't directly expose num_qubits.
        // Read it back from the ising model.
        unsafe { (*self.ising.ptr).num_qubits }
    }

    /// Run the QAOA optimisation loop to convergence.
    pub fn solve(&mut self) -> Result<QaoaResult> {
        let r: qaoa_result_t = unsafe { qaoa_solve(self.solver) };
        let n = r.num_layers;
        let gamma = if r.optimal_gamma.is_null() || n == 0 {
            Vec::new()
        } else {
            unsafe { std::slice::from_raw_parts(r.optimal_gamma, n).to_vec() }
        };
        let beta = if r.optimal_beta.is_null() || n == 0 {
            Vec::new()
        } else {
            unsafe { std::slice::from_raw_parts(r.optimal_beta, n).to_vec() }
        };
        Ok(QaoaResult {
            best_bitstring: r.best_bitstring,
            best_energy: r.best_energy,
            num_iterations: r.num_iterations,
            converged: r.converged != 0,
            approximation_ratio: r.approximation_ratio,
            optimal_gamma: gamma,
            optimal_beta: beta,
            num_layers: n,
            total_measurements: r.total_measurements,
        })
    }

    /// Evaluate `<H_C>(gamma, beta)` at the given angles without
    /// running the optimiser.  Useful for plotting the QAOA energy
    /// landscape.  Lengths of `gamma` and `beta` must equal the
    /// solver's `num_layers`.
    pub fn compute_expectation(&self, gamma: &[f64], beta: &[f64]) -> f64 {
        unsafe {
            qaoa_compute_expectation(
                self.solver,
                gamma.as_ptr(),
                beta.as_ptr(),
            )
        }
    }
}

impl Drop for QaoaSolver {
    fn drop(&mut self) {
        if !self.solver.is_null() {
            unsafe { qaoa_solver_free(self.solver) };
            self.solver = ptr::null_mut();
        }
        // ising + _entropy drop in field order.
    }
}

// QaoaSolver currently isn't Send/Sync because the C-side solver
// holds raw pointers into the ising model; same restriction as the
// other algorithm wrappers in this crate.

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn graph_rejects_invalid_edge_index() {
        let bad = Graph::new(3, &[(0, 5, 1.0)]); // vertex 5 out of range
        assert!(bad.is_err());
    }

    #[test]
    fn ising_evaluate_returns_finite_energy_on_triangle() {
        let g = Graph::new(3, &[(0, 1, 1.0), (1, 2, 1.0), (2, 0, 1.0)]).unwrap();
        let ising = IsingModel::from_maxcut(&g).unwrap();
        let e_000 = ising.evaluate(0b000);
        let e_101 = ising.evaluate(0b101);
        // Triangle MaxCut: all-aligned cuts 0 edges (E maximal),
        // one-flipped cuts 2 edges (E minimal).
        assert!(
            e_000 > e_101,
            "all-zeros (no cut, E={e_000}) should have higher energy than one-flipped (cut, E={e_101})"
        );
        assert!(e_101.is_finite());
    }

    #[test]
    fn qaoa_solver_runs_on_triangle_maxcut() {
        let g = Graph::new(
            3,
            &[(0, 1, 1.0), (1, 2, 1.0), (2, 0, 1.0)],
        )
        .unwrap();
        let ising = IsingModel::from_maxcut(&g).unwrap();
        let mut solver = QaoaSolver::new(ising, /*p=*/1).unwrap();
        let result = solver.solve().unwrap();
        assert_eq!(result.num_layers, 1);
        assert_eq!(result.optimal_gamma.len(), 1);
        assert_eq!(result.optimal_beta.len(), 1);
        assert!(result.best_energy.is_finite());
    }

    #[test]
    fn compute_expectation_returns_finite() {
        let g = Graph::new(2, &[(0, 1, 1.0)]).unwrap();
        let ising = IsingModel::from_maxcut(&g).unwrap();
        let solver = QaoaSolver::new(ising, /*p=*/1).unwrap();
        let e = solver.compute_expectation(&[0.3], &[0.5]);
        assert!(e.is_finite());
    }
}
